### ISSUE TITLE
discuss version managers in getting started

### ISSUE THREAD

**Issue by ashleygwilliams** [npm/docs#377]
because it's a best practice and also helps avoid the permissions problem. thoughts?

**Comment by aredridel** [comment]
I don't know that it's a best practice without context of the problem being solved, but discussion would be awesome to see. (I don't find I need one at all using homebrew, and I use per-project node versions by installing them as modules. Not quite ready for general consumption to do what I do, but there's other ways to solve some of these problems)

**Comment by ashleygwilliams** [comment]
@aredridel most people really dislike needing to use sudo to install things globally but are also made nervous by the permissions fix options. using nvm means no worries because it puts node_modules in your home directory..

**Comment by aredridel** [comment]
Yeah. homebrew's recommendations are to have `/usr/local` owned by people, and so you don't need sudo either.

**Comment by ashleygwilliams** [comment]
i guess i am wondering: do you think it's worth drafting up a new version of the getting started page so we have something concrete to work with? or is this not a good idea? (very open to people's thoughts on this)

**Comment by aredridel** [comment]
Considering just how prevalent `nvm` is, and how much people seem to need this, heck yes

**Comment by wilmoore** [comment]
> because it's a best practice and also helps avoid the permissions problem. thoughts?

I wouldn't say "best practice", in fact, some might even go as far as to call such tools overkill. I personally go back and forth depending on the situation.

I think I see where you are trying to go though. What you might be trying to convey is:

> Don't install stuff with sudo

I 100% agree with that.

...I just don't want to go with **"Just use nvm"** since that __might__ not be the final answer.

0. Many people know how to avoid using `sudo` without having to install a version manager; for these people, a version manager is overkill.
0. For those that do not know how to avoid `sudo`, perhaps we can focus on bringing them up to speed?
0. `nvm` has some fairly annoying opinions about how you *MUST* configure or not configure npm (i.e. `NPM_CONFIG_PREFIX`). Deal breaker for some people (self including). It breaks the spirit of being able to [configure npm](https://docs.npmjs.com/misc/scripts#best-practices) (see the 3rd bullet point).
  - https://github.com/creationix/nvm/issues/692#issuecomment-81402312

I don't want to necessarily be singling out `nvm`; because it isn't really just about `nvm`. The thing to convey is that there is nothing really special about a version manager regarding `sudo` other than it defaults to installing versions into the user's home directory. This is something the user can do on their own very easily. I personally install __everything__ into a subdirectory of my home. The real value of something __like__ `nvm` is the fact that it will install a `node` binary to a place that is guaranteed to be owned by the person running the install script.

Here is how to do with just `curl`:

```
mkdir -p $NODE_HOME/versions/node-v4.2.1
curl -sSL# http://nodejs.org/dist/latest/node-v4.2.1-linux-x64.tar.gz | tar xvz --strip 1 -C $NODE_HOME/versions/node-v4.2.1
export PATH="$NODE_HOME/versions/node-v4.2.1/bin:$PATH"
```

> NOTE: Make sure `$NODE_HOME` is somewhere under `$HOME`.

But an even easier way is to install node via homebrew. In fact, I would recommend to not let homebrew take over `/usr/local`; that is a formula :smile: for disaster. Installing homebrew into `$USER/.homebrew` makes many homebrew issues simply go away. The maintainers will tell you otherwise, but they really just mean there are a couple obscure programs that barf because they are hardcoded for `/usr/local`.

I personally feel like version managers can potentially introduce unnecessary complexity to a problem many people do not have. This is coming from someone that has written and contributed to version managers for multiple languages (i.e. not just node).

Today, I install a single node version (the latest) and that takes care of 99% of things. For the 1% (and 1% is generous), it is super simple to use something like direnv to switch versions for a specific project; however, I hardly ever do that...I have practiced doing it in case it ever comes up, but I don't need it yet. Most projects will not break if you are using the latest node. Furthermore, my work is increasingly being done in containers (i.e. docker), so that 1% is really more like 0%.

For me, working on many client projects, my most pressing issue has not been switching node versions. It has been more about this client or that client having a private npm registry, some using things like artifactory, some using enterprise npm, some using X. Keeping those credentials out of version control but making sure everything is configured such that when I'm in their private project, npm is talking to the correct registry and using the correct credentials.

I'm not saying, nobody should be using a version manager and everyone should be using containers. I'm just saying, maybe we should not assume everybody should be using a version manager as it is might not be as big a win as some would have us think.

**Comment by ashleygwilliams** [comment]
hi @wilmoore thanks so much so this *super* thoughtful comment.

i 100% agree. npm doesn't want to come out as supporting any particular technology. for certain people, certain tools will be better solutions. i would hope to present several different things!

i really love all the commentary here and will definitely use it in the doc! thanks again.

**Comment by wilmoore** [comment]
Hi @ashleygwilliams,

No problem. It's a thing that has been on my mind for a while and I had been wanting to write about it anyway. Your comment inspired me to barf this all out :smile: 

I think the cool thing about showing the `curl` instructions is that it is kinda cross-platform in that it works on linux and sort of works on windows as well if you just substitute `export PATH=...` for that weird thing in windows where you add environment variables via some super tiny popup window :laughing:
